### PR TITLE
GitHubReleaseTask - Initial set of changes - ChangeLog Enhancements

### PR DESCRIPTION
**Pull request by TheLayman**
Previously, ChangeLog used to compute the changes between the commit at which release is being created and the commit of the previous full release (non-draft, non-prerelease). 

With this PR, adding a compareWith attribute with three options: lastFullRelease, lastRelease and lastReleaseByTag. By default, the option selected is lastFullRelease. If the user selects lastReleaseByTag, 
he has to specify releaseTag attribute. [Tag or regex]

lastFullRelease : Compares the difference between current commit(at which the release is being created) and latest full release...... Existing Behavior
lastRelease:  Compares the difference between current commit(at which the release is being created) and latest non-draft release (Can be prerelease too).
lastReleaseByTag:  Compares the difference between current commit(at which the release is being created) and latest release matching the tag pattern specified in releaseTag.

### CODE REVIEW

File: Tasks/GitHubReleaseV0/main.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import tl = require("azure-pipelines-task-lib/task");
 import path = require("path");
 import { Action } from "./operations/Action";
-import { Utility, ActionType, Delimiters, TagSelectionMode} from "./operations/Utility";
+import { Utility, ActionType, Delimiters, ChangeLogStartCommit } from './operations/Utility';
```
Comment (shyama1208) [inline Tasks/GitHubReleaseV0/main.ts]: nit: double quotes

File: Tasks/GitHubReleaseV0/task.json [hunk 2]
```diff
@@ -187,6 +187,28 @@
             "required": false,
             "visibleRule": "action = create || action = edit",
             "helpMarkDown": "If set to `true`, a list of changes (commits and issues) between this and the last published release will be generated and appended to the release notes."
+        },
```
Comment (shyama1208) [inline Tasks/GitHubReleaseV0/task.json]: change either patch or minor version of the task.

Comment (TheLayman) [reply Tasks/GitHubReleaseV0/task.json]: Version 0.155 is not live yet. Hence didn't change the version.

File: Tasks/GitHubReleaseV0/Strings/resources.resjson/en-US/resources.resjson [hunk 3]
```diff
@@ -35,6 +35,10 @@
   "loc.input.help.isPreRelease": "Indicate whether the release should be marked as a pre-release.",
   "loc.input.label.addChangeLog": "Add changelog",
   "loc.input.help.addChangeLog": "If set to `true`, a list of changes (commits and issues) between this and the last published release will be generated and appended to the release notes.",
+  "loc.input.label.compareWith": "Compare With",
```
Comment (nikhilexe) [inline Tasks/GitHubReleaseV0/Strings/resources.resjson/en-US/resources.resjson]: sentence casing: Compare with

File: Tasks/GitHubReleaseV0/Strings/resources.resjson/en-US/resources.resjson [hunk 4]
```diff
@@ -35,6 +35,10 @@
   "loc.input.help.isPreRelease": "Indicate whether the release should be marked as a pre-release.",
   "loc.input.label.addChangeLog": "Add changelog",
   "loc.input.help.addChangeLog": "If set to `true`, a list of changes (commits and issues) between this and the last published release will be generated and appended to the release notes.",
+  "loc.input.label.compareWith": "Compare with",
+  "loc.input.help.compareWith": "Indicate which release we should compare with to generate the changelog.",
+  "loc.input.label.releaseTag": "Release Tag",
```
Comment (prativen) [inline Tasks/GitHubReleaseV0/Strings/resources.resjson/en-US/resources.resjson]: Need to review the label. We should get PM input here

File: Tasks/GitHubReleaseV0/Strings/resources.resjson/en-US/resources.resjson [hunk 5]
```diff
@@ -81,7 +85,10 @@
   "loc.messages.CommitDiffBehind": "Cannot compute the changes as the provided target commit is older than the commit of the last published release.",
   "loc.messages.CommitDiffEqual": "No changes were found. The provided target commit is the same as the commit of the last published release.",
   "loc.messages.FetchLatestPublishRelease": "Fetching the latest published release...",
+  "loc.messages.FetchLatestNonDraftRelease": "Fetching the latest non-draft release...",
+  "loc.messages.FetchLastReleaseByTag": "Fetching the latest release matching the given tag",
   "loc.messages.FetchLatestPublishReleaseSuccess": "Found the latest published release: %s",
+  "loc.messages.FetchMatchingReleaseSuccess": "Found the latest release matching the given criteria.",
```
Comment (prativen) [inline Tasks/GitHubReleaseV0/Strings/resources.resjson/en-US/resources.resjson]: dont we need a found message for latest non-draft release?

File: Tasks/GitHubReleaseV0/operations/Constants.ts [hunk 6]
```diff
@@ -17,4 +17,6 @@ export class Inputs {
     public static readonly addChangeLog = "addChangeLog";
     public static readonly deleteExistingAssets = "deleteExistingAssets";
     public static readonly tagPattern = "tagPattern";
+    public static readonly compareWith = "compareWith";
+    public static readonly releaseTag = "releaseTag";
```
Comment (prativen) [inline Tasks/GitHubReleaseV0/operations/Constants.ts]: we should review these input literals

File: Tasks/GitHubReleaseV0/operations/Utility.ts [hunk 7]
```diff
@@ -221,6 +221,13 @@ export class Utility {
         }
     }
 
+    public static validateStartCommitSpecification(compareWith: string) {
+        if (compareWith !== startCommitSpecification.lastFullRelease && compareWith !== startCommitSpecification.lastRelease
```
Comment (prativen) [inline Tasks/GitHubReleaseV0/operations/Utility.ts]: case sensitivity?

File: Tasks/GitHubReleaseV0/operations/Utility.ts [hunk 8]
```diff
@@ -261,6 +268,18 @@ export class AssetUploadMode {
     public static readonly replace = "replace";
 }
 
+export class startCommitSpecification {
```
Comment (prativen) [inline Tasks/GitHubReleaseV0/operations/Utility.ts]: should this be changeLogStartCommitSpecification to be more specific

File: Tasks/GitHubReleaseV0/operations/Utility.ts [hunk 9]
```diff
@@ -221,6 +221,13 @@ export class Utility {
         }
     }
 
+    public static validateStartCommitSpecification(compareWith: string) {
+        if (compareWith !== startCommitSpecification.lastFullRelease && compareWith !== startCommitSpecification.lastRelease
+            && compareWith != startCommitSpecification.lastReleaseByTag) {
+            throw new Error(tl.loc("InvalidCompareWithAttribute", compareWith));
```
Comment (shyama1208) [inline Tasks/GitHubReleaseV0/operations/Utility.ts]: does this surface the errors in logs?

File: Tasks/GitHubReleaseV0/main.ts [hunk 10]
```diff
@@ -102,9 +102,12 @@ class Main {
         const releaseNoteInput = tl.getInput(Inputs.releaseNotes);
         const showChangeLog: boolean = tl.getBoolInput(Inputs.addChangeLog);
 
+        const compareWithRelease = tl.getInput(Inputs.compareWith);
+        showChangeLog && Utility.validateStartCommitSpecification(compareWithRelease);
+        const releaseTag = tl.getInput(Inputs.releaseTag);
```
Comment (prativen) [inline Tasks/GitHubReleaseV0/main.ts]: this input may not be set in all conditions. what does this return in that case?

Comment (prativen) [reply Tasks/GitHubReleaseV0/main.ts]: you should explicitly read for releaseTag only when comparewithRelease is the tag option else set it to empty

File: Tasks/GitHubReleaseV0/operations/ChangeLog.ts [hunk 11]
```diff
@@ -12,37 +12,124 @@ export class ChangeLog {
...
-        if (latestReleaseResponse.statusCode === 200 || latestReleaseResponse.statusCode === 404) {
-            // Get the curent commit.
-            let endCommitSha: string = await new Helper().getCommitShaFromTarget(githubEndpointToken, repositoryName, target);
 
+        // Get the curent commit.
+        let endCommitSha: string = await new Helper().getCommitShaFromTarget(githubEndpointToken, repositoryName, target);
+        //Get the start commit.
+        let startCommitSha: string = await this.getStartCommitSha(githubEndpointToken, repositoryName, endCommitSha, top,compareWithRelease, releaseTag);
+        // Compare the diff between 2 commits.
+        console.log(tl.loc("FetchCommitDiff"));
```
Comment (prativen) [inline Tasks/GitHubReleaseV0/operations/ChangeLog.ts]: we might want to log start and end commit in tl.debug mode?

File: Tasks/GitHubReleaseV0/operations/Utility.ts (hunk 9 above)
Comment (shyama1208) [reply Tasks/GitHubReleaseV0/operations/Utility.ts]: yes

File: Tasks/GitHubReleaseV0/operations/ChangeLog.ts [hunk 12]
```diff
@@ -53,84 +140,84 @@ export class ChangeLog {
...
+    /**
+     * Returns latest release satisfying the given comparer.
+     * @param githubEndpointToken 
+     * @param repositoryName 
+     * @param comparer 
+     */
+    public async getLastReleaseTag(githubEndpointToken: string, repositoryName: string, comparer:(release: any)=> boolean): Promise<string> {
+        let release = new Release();
+        
+        // Fetching all releases in the repository.
```
Comment (prativen) [inline Tasks/GitHubReleaseV0/operations/ChangeLog.ts]: in descending order

File: Tasks/GitHubReleaseV0/operations/ChangeLog.ts [hunk 13]
```diff
@@ -53,84 +140,84 @@ export class ChangeLog {
...
+                links = Utility.parseHTTPHeaderLink(releasesResponse.headers[GitHubAttributes.link]);
+
+                // Calling the next page if it exists
+                if (links && links[GitHubAttributes.next]) {
+                    let paginatedResponse = await release.getPaginatedResult(githubEndpointToken, links[GitHubAttributes.next]);
+                    releasesResponse = paginatedResponse;
+                    continue;
+                }
+            }
+            //If status code is 404 or there are no releases satisfying the constraints return null.
```
Comment (prativen) [inline Tasks/GitHubReleaseV0/operations/ChangeLog.ts]: put a log to indicate the scenario of 404 or not matching releases

File: Tasks/GitHubReleaseV0/operations/ChangeLog.ts [hunk 14]
```diff
@@ -53,84 +140,84 @@ export class ChangeLog {
...
-            else{
-                tl.error(tl.loc("FetchCommitDiffError"));
-                throw new Error(commitsListResponse.body[GitHubAttributes.message]);
-            }
+        let initialTag = await this.getLastReleaseTag(githubEndpointToken, repositoryName, comparer);
+
+        //If no such release exists, get the start commit
+        //else get the commit for that tag.
+        if (!initialTag) {
+            console.log(tl.loc("FetchInitialCommit"));
```
Comment (prativen) [inline Tasks/GitHubReleaseV0/operations/ChangeLog.ts]: log the condition failure (no releases or no matching draft release..)

**Pull request merged**